src/main/java/com/devonfw/ide/sonarqube/common/impl/check/namingconvention/DevonNamingConventionClassExtendsClassDaoCheck.java

```diff
@@ -0,0 +1,77 @@
+package com.devonfw.ide.sonarqube.common.impl.check.namingconvention;
+
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.ClassTree;
+
+/**
+ * {@link DevonNamingConventionClassExtendsClassCheck} verifying that non-abstract classes extending AbstractDao are
+ * following the devonfw naming convention by ending with DaoImpl.
+ */
+@Rule(key = "Devon4j:N2", name = "Devon naming conventions of inheriting classes (DaoImpl).", //
+    description = "Verify that non-abstract Classes extending AbstractDao shall end with DaoImpl. In addition, classes "
+        + "that inherit from a class with the suffix Dao must also have Dao as their suffix.", //
+    priority = Priority.CRITICAL, tags = { "architecture-violation" })
+public class DevonNamingConventionClassExtendsClassDaoCheck extends DevonNamingConventionClassExtendsClassCheck {
+
+  private static final String DESIRED_SUPERCLASS_NAME = "AbstractDao";
+
+  /**
+   * Calls super constructor to compile the pattern classSuffiRegEx
+   */
+  public DevonNamingConventionClassExtendsClassDaoCheck() {
+
+    super(".*DaoImpl");
+  }
+
+  /**
+   * Method called after parsing and semantic analysis has been done on file.
+   *
+   * @param context Context of analysis containing the parsed tree.
+   */
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+
+    ClassTree tree = getTreeInstance(context);
+    this.className = tree.simpleName().name();
+    this.superClassName = tree.superClass().toString();
```

Comment:
For the two specific rules this block is duplicated and also part of the checks seems to be kind of redundant and also not so easy to read (`if (!A && !B) || (A && B && !abstract) { return;}`).
It is all correct and already improved very well, but I was thinking this over and came to the following suggestion for the final improvement:
* In `DevonNamingConventionClassExtendsClassCheck.scanFile` extract the `this.classSuffixRegEx.matcher(this.superClassName).matches()` to a protected method `isSuperClassMatching`
* This method can be overriden in the 2 specific rules to also check the specific super-class name as well and otherwise delegate to default behaviour (`return super.isSuperClassMatching()`).
* Then move the second condition and the if body of `DevonNamingConventionClassExtendsClassCheck.scanFile` to yet another protected method `checkClass`
* Again the 2 specific rules can override this `checkClass` method and add their extra stuff such as
```
if (isAbstract(tree)) {
  return
}
```
or the extra check for `desiredSuperInterfaceName`. Please note that you pass parameters to the protected methods as needed for the specific cases.

Does that make sense?
Sorry, for the long way to get this perfect. But IMHO this would give it the final improvement.

Reply:
I extracted the methods into the super class, but I'm not completely sure I understood which logic to put in the overridden methods. 
Using the default behavior for the two special classes would not work IMO as the logic from the super classes scanFile() can not be used for those. The parameters that are given here to the super constructor deviate from the parameters given by the "normal" classes.
I'm sure we can resolve this when we talk on Friday.

Reply:
Thanks for the quick reaction and already the updates.
Sure we can easily resolve this on Friday and together we will quickly be able to solve it. Feel free to pick another issue till then...